rs/tests/crypto/ingress_verification_test.rs

```diff
@@ -0,0 +1,634 @@
+/* tag::catalog[]
+end::catalog[] */
+use anyhow::Result;
+use ic_agent::Identity;
+use ic_agent::export::Principal;
+use ic_crypto_test_utils_reproducible_rng::reproducible_rng;
+use ic_registry_subnet_type::SubnetType;
+use ic_system_test_driver::driver::group::{SystemTestGroup, SystemTestSubGroup};
+use ic_system_test_driver::driver::ic::{InternetComputer, Subnet};
+use ic_system_test_driver::driver::test_env::TestEnv;
+use ic_system_test_driver::driver::test_env_api::{GetFirstHealthyNodeSnapshot, HasPublicApiUrl};
+use ic_system_test_driver::systest;
+use ic_system_test_driver::util::{
+    UniversalCanister, block_on, expiry_time, sign_query, sign_update,
+};
+use ic_types::crypto::SignedBytesWithoutDomainSeparator;
+use ic_types::messages::{
+    Blob, Delegation, HttpCallContent, HttpCanisterUpdate, HttpQueryContent, HttpRequestEnvelope,
+    HttpUserQuery, SignedDelegation,
+};
+use ic_types::{CanisterId, PrincipalId, Time};
+use rand::{CryptoRng, Rng};
+use reqwest::{StatusCode, Url};
+use slog::{debug, info};
+
+fn main() -> Result<()> {
+    SystemTestGroup::new()
+        .with_setup(setup)
+        .add_parallel(
+            SystemTestSubGroup::new()
+                .add_test(systest!(requests_with_delegations; 2))
+                .add_test(systest!(requests_with_delegations; 3))
+                .add_test(systest!(requests_with_delegations_with_targets; 2))
+                .add_test(systest!(requests_with_delegations_with_targets; 3))
+                .add_test(systest!(requests_with_delegation_loop; 2))
+                .add_test(systest!(requests_with_delegation_loop; 3)),
+        )
+        .execute_from_args()?;
+    Ok(())
+}
+
+pub fn setup(env: TestEnv) {
+    InternetComputer::new()
+        .add_subnet(Subnet::fast_single_node(SubnetType::System))
+        .add_subnet(Subnet::fast_single_node(SubnetType::Application))
+        .setup_and_start(&env)
+        .expect("failed to setup IC under test");
+}
+
+#[derive(Clone, Debug)]
+struct TestInformation {
+    api_ver: usize,
+    url: Url,
+    canister_id: Principal,
+}
+
+#[derive(Copy, Clone, Debug)]
+enum GenericIdentityType {
+    Ed25519,
+    EcdsaSecp256k1,
+    EcdsaSecp256r1,
+    // TODO webauthn RSA
+    // TODO webauthn EC
+}
+
+impl GenericIdentityType {
+    fn random<R: Rng + CryptoRng>(rng: &mut R) -> Self {
+        match rng.r#gen::<usize>() % 3 {
+            0 => Self::EcdsaSecp256k1,
+            1 => Self::EcdsaSecp256r1,
+            _ => Self::Ed25519,
+        }
+    }
+}
+
+#[derive(Clone)]
+enum GenericIdentityInner {
+    K256(ic_secp256k1::PrivateKey),
+    P256(ic_secp256r1::PrivateKey),
+    Ed25519(ic_ed25519::PrivateKey),
+}
+
+#[derive(Clone)]
+struct GenericIdentity {
+    inner: GenericIdentityInner,
+    public_key: Vec<u8>,
```

Comment:
nit: how about being explicit and calling this (and the getter) `public_key_der`?